src/components/js-tracer/index.js

```diff
@@ -0,0 +1,97 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// @flow
+import * as React from 'react';
+import explicitConnect from '../../utils/connect';
+import JsTracerChart from './Chart';
+import JsTracerSettings from './Settings';
+import EmptyReasons from './EmptyReasons';
+
+import {
+  getProfile,
+  selectedThreadSelectors,
+} from '../../reducers/profile-view';
+import {
+  getShowJsTracerSummary,
+  getSelectedThreadIndex,
+} from '../../reducers/url-state';
+import { updatePreviewSelection } from '../../actions/profile-view';
+import { changeSelectedTab } from '../../actions/app';
+
+import type { Profile, JsTracerTable, ThreadIndex } from '../../types/profile';
+import type {
+  ExplicitConnectOptions,
+  ConnectedProps,
+} from '../../utils/connect';
+
+require('./index.css');
+
+type DispatchProps = {|
+  +updatePreviewSelection: typeof updatePreviewSelection,
+  +changeSelectedTab: typeof changeSelectedTab,
+|};
+
+type StateProps = {|
+  +profile: Profile,
+  +threadIndex: ThreadIndex,
+  +jsTracerTable: JsTracerTable | null,
+  +showJsTracerSummary: boolean,
+|};
+
+type Props = ConnectedProps<{||}, StateProps, DispatchProps>;
+
+class JsTracer extends React.PureComponent<Props> {
+  _rafGeneration: number = 0;
+
+  componentDidUpdate() {
+    const { changeSelectedTab, jsTracerTable } = this.props;
+    if (jsTracerTable === null) {
+      // If the user switches to another thread that doesn't have JS Tracer information,
+      // then switch to the calltree.
+      changeSelectedTab('calltree');
+    }
```

Comment:
This feels a bit "spaghetti". Can't we do this in the action creator `selectTrack`?